rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+- Feature Name: `error_type`
+- Start Date: 2024-11-18
+- RFC PR: [FuelLabs/sway-rfcs#43](https://github.com/FuelLabs/sway-rfcs/pull/43)
+- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/001)
+
+# Summary
+
+[summary]: #summary
+
+This RFC aims to standardize error handling in Sway so that there is a
+deliberate and clear way to produce error messages that are easily consumable by
+language tools while keeping error strings out of on-chain artifacts.
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently, reverting a sway program only produces an opaque error code.
+Sometimes this code is unusual enough that with the source code of the contract
+it is possible to figure out what is going on, but a lot of the time it's a
+meaningless number that only means that some failure happened somewhere.
+
+It's not reasonable to expect people to be able to debug sway programs in these
+conditions, so we need to improve this.
+
+We want irrecoverable errors to carry user defined error messages and source
+information about where the error was invoked. We want those error types to be
+something that can also be used for recoverable errors.
+
+To this end we want to introduce a mechanism that makes error codes an
+understandable part of the ABI, and language features that allow the user to
+define error messages, error types and invoke those errors with the appropriate
+information being generated.
+
+We also want recoverable errors to connect to this mechanism so that all error
+handling in a given Sway program can go through one or a small set of user
+defined types and error strings don't affect the final bytecode size.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Error types
+
+Error types are special enums that allow developers to signal that each variant
+is associated with an error message.
+
+An error type implements the `Error` trait, but it is not recommended that you
+do so manually.
+
+Instead, to create a new error type, use the following syntax and annotations on
+an enum:
+
+```sway
+#[error_type]
+enum MyError {
+  #[error("error A")]
+  A: (),
+  #[error("error B")]
```

Comment:
Will there be some way to include the error parameters in the error message? Or, will the parameters only be surfaced to the user in the log receipts?

We could support the same syntax as `thiserror` like so:

```sway
  #[error("error B first param: {0}, second param: {1}")]
  B: (u64, u8),
```

Reply:
There isn't, but this is listed in the possible extensions.

The reason I didn't include this directly here is that we don't actually have format strings in Sway at all yet. And you would want the syntax for this to be the standard one you use everywhere.